Update EIP-7002: Add system call gas aspects and directive for handling failures

somnathb1 opened the pull request.

jochem-brouwer commented on EIPS/eip-7002.md:
```diff
@@ -706,6 +706,10 @@ Note: the above code reverts if the fee is too high, the fee can change signific
 
+### System Call failure
+
+Although the likelihood of a failed system call to SYSTEM_ADDRESS is extremely low, the behavior in such cases is well-defined: the block is marked as invalid. However, if the failure results from processing a transaction within the block, the public mempool may still retain the transaction even after the block is invalidated. This can result in the offending transaction being included again, pot ...
```
Same here, `SYSTEM_ADDRESS` does not (should not) have code, but `WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS` has the relevant code here.

somnathb1 commented on EIPS/eip-7002.md:
````diff
@@ -226,7 +226,7 @@ def reset_withdrawal_requests_count():
 ```
 
-Each withdrawal request must appear in the EIP-7685 requests list in the order they are returned from `dequeue_withdrawal_requests()`.
+Each withdrawal request must appear in the EIP-7685 requests list in the order they are returned from `dequeue_withdrawal_requests()`. If, however, at the end of processing an execution block the System Call to the `SYSTEM_ADDRESS` contract fails, the block must be considered invalid.
````
Yes, thanks!

mkalinin commented:
Looks good to me in general. A question: what are potential ways for a system call to fail? If the call is anyhow parametrized then we must specify those parameters in the EIP to make EL clients agree on them and avoid potential splits

somnathb1 commented:
Currently it's not parameterized. System call can fail if
- The contract is bad or a transaction triggered an unexpected behaviour
- The contract/syscall runs out of gas (30M, say)
- Erroneous/Differing EL implementation - currently we don't guard against this because it's trivial

> Looks good to me in general. A question: what are potential ways for a system call to fail? If the call is anyhow parametrized then we must specify those parameters in the EIP to make EL clients agree on them and avoid potential splits

mkalinin commented:
> The contract/syscall runs out of gas (30M, say)

30M sounds like a parameter. I think it worth adding it to the EIP

somnathb1 commented:
> > The contract/syscall runs out of gas (30M, say)
> 
> 30M sounds like a parameter. I think it worth adding it to the EIP

It's only the chain's block gas limit. And yes, we should specify what the System Call looks like in terms of gas limit.

mkalinin commented:
> It's only the chain's block gas limit. And yes, we should specify what the System Call looks like in terms of gas limit.

Right, i suppose a system call doesn’t count into gas used by block transactions? Why should it be limited by 30M?

somnathb1 commented:
> Right, i suppose a system call doesn’t count into gas used by block transactions? Why should it be limited by 30M?
- An upper ceiling to avoid block delay
- A common theme in system calls
- We can get rid of that, but it may conflict with how it's implemented in practice in EL clients

mkalinin commented:
> We can get rid of that, but it may conflict with how it's implemented in practice in EL clients

I think we should rather make it explicit in all EIPs with gas limited system calls

somnathb1 commented:
> I think we should rather make it explicit in all EIPs with gas limited system calls

Added the definitions, and also added invalidate condition for when there is nothing deployed at the `WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS`. This has the pro that if clients have differing configs, this will cause a stir. But the downside is a chain cannot operate without this.

I.m.o the post-pectra readiness always involves deploying this address. And absence of the contract that has associated system calls should not be considered a valid scenario or be "failed silently" with

jochem-brouwer commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
+* Gas consumed by this call does not count against the block’s overall gas usage.
+* Both the gas limit assigned to the call and the gas consumed **MUST** be excluded from any checks against the block’s gas limit.
+* The call does not follow [EIP-1559](./eip-1559.md) fee burn semantics—value **MUST NOT** be transferred as part of this call.
+* If there is no code at `WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS`, the corresponding block **MUST** be marked invalid.
```
The problem here is (I think this was also raised by @marioevz) that if there is no code at the fork block, then the block is thus invalid (at the fork block). Since now every block which is being produced is invalid, it is impossible to deploy the withdrawal system contract.

The suggestion which I would take here is to just follow the path as in the normal call, except there is no code at the address. Therefore, the returndata is the empty bytes, and therefore also the request data itself is empty, so it will not alter the requestsHash but it will ensure that the block is valid and it is possible to deploy the withdrawals contract.

somnathb1 replied:
Pectra (or EIP-7002 specifically) should not be activated before being Pectra (or, here, EIP-7002) ready. This is not a problem, it's deliberate. 

As i highlighted on the call, one way to then achieve chain continuation is pushing the timestamp of the fork block, deploy the contract and continue. The chain would stop on the pre-fork block, ready for this change

jochem-brouwer replied:
Right, sorry, you indeed raised this point during the call, I forgot.

I'm not sure if this process works. So, first we set fork at time X but do not recognize the system contract is not deployed. Then, at time X all blocks produced are invalid. So now we re-coordinate a fork time later for clients, we deploy this new fork time (how does this work in practice? Nodes will get un-stuck not all at the same time, but they will get un-stuck if they now update the client to the later fork time X). So now we have a chain which rolls forward, we deploy the system contract, and then we succesfully fork (this could work but I'm not sure what would happen if clients get un-stuck piece-by-piece? Is that a problem?)

jochem-brouwer replied:
Actually I think it would be nice if you put this point somewhere, since this is a super good point, it only allows the chain to move forward once the contract is deployed and this must happen at the fork block if it isn't there (I did not get this directly from this point, I assumed if the chain gets stuck it would just be doomed unless the fork would be rescheduled, which is not the case)

mkalinin commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
+
+Additionally, the system call and subsequent block processing **MUST** conform to the following:
+
+* The call has a dedicated gas limit of `30_000_000`.
```
```suggestion
* The call **MUST** have a dedicated gas limit of `30_000_000`.
```

mkalinin commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
+Additionally, the system call and subsequent block processing **MUST** conform to the following:
+
+* The call has a dedicated gas limit of `30_000_000`.
+* Gas consumed by this call does not count against the block’s overall gas usage.
```
```suggestion
* Gas consumed by this call **MUST NOT** count against the block’s overall gas usage.
```

somnathb1 replied:
Already says  `**MUST** conform` above, this line is just descriptive.

mkalinin replied:
I would prefer to use MUST NOT instead of “does not". But is this is just my preference and otherwise it reads clear then i am happy to leave “does not” as is

somnathb1 replied:
We can leave "does not" as is, it's very clear and impossible to confuse with any hint of optionality in this context.

mkalinin commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
+* Both the gas limit assigned to the call and the gas consumed **MUST** be excluded from any checks against the block’s gas limit.
+* The call does not follow [EIP-1559](./eip-1559.md) fee burn semantics—value **MUST NOT** be transferred as part of this call.
+* If there is no code at `WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS`, the corresponding block **MUST** be marked invalid.
+* If the call to the contract fails or returns an error, the block **MUST** be invalidated
```
```suggestion
* If the call to the contract fails or returns an error, the block **MUST** be invalidated.
```

mkalinin commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
+* The call has a dedicated gas limit of `30_000_000`.
+* Gas consumed by this call does not count against the block’s overall gas usage.
+* Both the gas limit assigned to the call and the gas consumed **MUST** be excluded from any checks against the block’s gas limit.
+* The call does not follow [EIP-1559](./eip-1559.md) fee burn semantics—value **MUST NOT** be transferred as part of this call.
```
```suggestion
* The call does not follow [EIP-1559](./eip-1559.md) fee burn semantics — value **MUST NOT** be transferred as part of this call.
```

lightclient commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
-Specifically, the functionality is defined in pseudocode as the function `read_withdrawal_requests()`:
+Each withdrawal request must appear in the EIP-7685 requests list in the exact order returned by `dequeue_withdrawal_requests()`.
+
+Additionally, the system call and subsequent block processing **MUST** conform to the following:
```
What do you mean by "subsequent block processing"?

lightclient replied:
also you can just copy to text from 4788 so the behavior and style of the eip is identical:

https://github.com/ethereum/EIPs/blob/c6bc36304bdef3ddce9ede45f0c7cb5901bb7e07/EIPS/eip-4788.md?plain=1#L238-L243

somnathb1 replied:
> What do you mean by "subsequent block processing"?

Remaining part of processing of the block in which the call is made. There are post block generation and validation checks.

If it's not clear, because of the word "subsequent", we can say "and the corresponding block processing"

somnathb1 replied:
> also you can just copy to text from 4788 so the behavior and style of the eip is identical:

It is mostly copied from there, yes, with some changes

lightclient commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
 ##### System Call
 
-At the end of processing any execution block starting from the `FORK_BLOCK` (i.e. after processing all transactions and after performing the block body withdrawal requests validations), call the contract as `SYSTEM_ADDRESS` and perform the following:
+At the end of processing any execution block starting from the `FORK_BLOCK` (i.e. after processing all transactions and after performing the block body withdrawal requests validations), the implementation **MUST** invoke a call to the contract at `WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS` as `SYSTEM_ADDRESS` with no calldata. This invocation triggers the following behavior:
```
```suggestion
At the end of processing any execution block starting from the `FORK_BLOCK` (i.e. after processing all transactions and after performing the block body withdrawal requests validations), call `WITHDRAWAL_REQUEST_PREDEPLOY_ADDRESS` as `SYSTEM_ADDRESS` with no calldata. The invocation triggers the following:
```

Not really using RFC 2119 vocabulary in this EIP so should keep the style the same.

somnathb1 replied:
That's true, but for the sake of additional clarification, it is okay. Do you have a stronger reason to fall back to the more informal language here?

somnathb1 replied:
Many of the `**MUST**`'s were also suggested by @mkalinin . I am okay with either. Just want to be as clear as possible. If it is indicated that less stricter language is also as clear, we can go for that

lightclient reviewed (commented):
change generally LGTM, left some nits

do we also need to update 7251?

somnathb1 commented:
> change generally LGTM, left some nits
> 
> do we also need to update 7251?

Planning to create a cc PR for 7251 after this one gets argued about and merged.

somnathb1 commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
-Specifically, the functionality is defined in pseudocode as the function `read_withdrawal_requests()`:
+Each withdrawal request must appear in the EIP-7685 requests list in the exact order returned by `dequeue_withdrawal_requests()`.
+
+Additionally, the system call and subsequent block processing **MUST** conform to the following:
```
```suggestion
Additionally, the system call and the processing of that block must conform to the following:
```

somnathb1 commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
+
+* The call has a dedicated gas limit of `30_000_000`.
+* Gas consumed by this call does not count against the block’s overall gas usage.
+* Both the gas limit assigned to the call and the gas consumed **MUST** be excluded from any checks against the block’s gas limit.
```
```suggestion
* Both the gas limit assigned to the call and the gas consumed is excluded from any checks against the block’s gas limit.
```

somnathb1 commented on EIPS/eip-7002.md:
```diff
@@ -144,13 +144,24 @@ When the input to the contract is length zero, interpret this as a get request f
+* The call has a dedicated gas limit of `30_000_000`.
+* Gas consumed by this call does not count against the block’s overall gas usage.
+* Both the gas limit assigned to the call and the gas consumed **MUST** be excluded from any checks against the block’s gas limit.
+* The call does not follow [EIP-1559](./eip-1559.md) fee burn semantics — value **MUST NOT** be transferred as part of this call.
```
```suggestion
* The call does not follow [EIP-1559](./eip-1559.md) fee burn semantics — no value to be transferred as part of this call.
```
